File: brdgme-go/love_letter/render.go
package love_letter

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/brdgme/brdgme/brdgme-go/brdgme"
	"github.com/brdgme/brdgme/brdgme-go/render"
)

func (g *Game) PubRender() string {
	return g.PlayerRender(-1)
}

func (g *Game) PlayerRender(pNum int) string {
	leader := brdgme.IntMax(g.PlayerPoints...)
	rows := []string{fmt.Sprintf(
		"The leader has {{b}}%d %s{{/b}}, the game will end at {{b}}%d points{{/b}}",
		leader,
		brdgme.Plural(leader, "point"),
		endScores[g.Players],
	), ""}

	if pNum >= 0 {
		if g.Eliminated[pNum] {
			rows = append(
				rows,
				render.Bold("You have been eliminated from this round"),
			)
		} else {
			rows = append(
				rows,
				render.Bold(fmt.Sprintf(
					"Your %s",
					brdgme.Plural(len(g.Hands[pNum]), "card"),
				)),
				strings.Join(RenderCards(g.Hands[pNum]), "   "),
			)
		}
	}

	playerTable := [][]render.Cell{
		{},
		{
			render.Cel(render.Bold("Player")),
			render.Cel(render.Bold("Pts"), render.Center),
			render.Cel(render.Bold("Status"), render.Center),
			render.Cel(render.Bold("Discards"), render.Center),
		},
	}
	for p := 0; p < g.Players; p++ {
		status := render.Markup("active", render.Green, true)
		if g.Eliminated[p] {
			status = render.Markup("eliminated", render.Grey, false)
		} else if g.Protected[p] {
			status = render.Markup("protected", render.Black, true)
		}
		playerTable = append(playerTable, []render.Cell{
			render.Cel(render.Player(p)),
			render.Cel(render.Bold(strconv.Itoa(g.PlayerPoints[p])), render.Center),
			render.Cel(status, render.Center),
			render.Cel(strings.Join(RenderCards(g.Discards[p]), "  ")),
		})
	}
	rows = append(rows,
		render.Table(playerTable, 0, 2),
		"",
		render.Bold(fmt.Sprintf("Cards remaining: %d", len(g.Deck))),
	)

	helpTable := [][]render.Cell{
		{},
		{},
		{
			render.Cel(render.Bold("Card")),
			render.Cel(render.Bold("#")),
			render.Cel(render.Bold("Description")),
		},
	}
	for c := Princess; c >= Guard; c-- {
		helpTable = append(helpTable, []render.Cell{
			render.Cel(RenderCard(c)),
			render.Cel(strconv.Itoa(brdgme.IntCount(c, Deck))),
			render.Cel(render.Fg(render.Grey, Cards[c].Text)),
		})
	}
	rows = append(rows, render.Table(helpTable, 0, 2))

	return render.Layout(rows)
}

func RenderCard(card int) string {
	return render.Markup(fmt.Sprintf(
		"%s (%d)",
		Cards[card].Name,
		Cards[card].Number,
	), Cards[card].Color, true)
}

func RenderCards(cards []int) []string {
	strs := make([]string, len(cards))
	for i, c := range cards {
		strs[i] = RenderCard(c)
	}
	return strs
}

File: rust/game/love-letter-2/src/render.rs
use brdgme_color::NamedColor;
use brdgme_game::Renderer;
use brdgme_markup::{Align as A, Node as N, Row, table_with_gap};

use crate::card::{Card, deck_count, princess_to_guard};
use crate::{PlayerState, PubState, plural};

/// Column spacing used by the Go `render.Table(cells, 0, 2)` calls in
/// `render.go` - two literal spacer cells between real columns.
const COL_SPACING: usize = 2;

fn render(pub_state: &PubState, player: Option<usize>, hand: Option<&[Card]>) -> Vec<N> {
    let mut rows: Vec<Row> = vec![];

    rows.push(vec![(
        A::Center,
        vec![
            N::text("The leader has "),
            N::Bold(vec![N::text(format!(
                "{} {}",
                pub_state.leader_points,
                plural(pub_state.leader_points, "point")
            ))]),
            N::text(", the game will end at "),
            N::Bold(vec![N::text(format!("{} points", pub_state.end_score))]),
        ],
    )]);
    rows.push(vec![]);

    if let Some(p) = player {
        if pub_state.eliminated.get(p).copied().unwrap_or(false) {
            rows.push(vec![(
                A::Center,
                vec![N::Bold(vec![N::text(
                    "You have been eliminated from this round",
                )])],
            )]);
        } else if let Some(h) = hand {
            rows.push(vec![(
                A::Center,
                vec![N::Bold(vec![N::text(format!(
                    "Your {}",
                    plural(h.len(), "card")
                ))])],
            )]);
            let mut hand_line: Vec<N> = vec![];
            for (i, &c) in h.iter().enumerate() {
                if i > 0 {
                    hand_line.push(N::text("   "));
                }
                hand_line.push(card(c));
            }
            rows.push(vec![(A::Center, hand_line)]);
        }
    }

    rows.push(vec![(
        A::Center,
        vec![table_with_gap(&player_table(pub_state), COL_SPACING)],
    )]);
    rows.push(vec![]);
    rows.push(vec![(
        A::Center,
        vec![N::Bold(vec![N::text(format!(
            "Cards remaining: {}",
            pub_state.deck_remaining
        ))])],
    )]);

    rows.push(vec![(
        A::Center,
        vec![table_with_gap(&help_table(), COL_SPACING)],
    )]);

    vec![N::Table(rows)]
}

fn player_table(pub_state: &PubState) -> Vec<Row> {
    let mut table: Vec<Row> = vec![
        vec![],
        vec![
            (A::Left, vec![N::Bold(vec![N::text("Player")])]),
            (A::Center, vec![N::Bold(vec![N::text("Pts")])]),
            (A::Center, vec![N::Bold(vec![N::text("Status")])]),
            (A::Center, vec![N::Bold(vec![N::text("Discards")])]),
        ],
    ];
    for p in 0..pub_state.players {
        let status = if pub_state.eliminated.get(p).copied().unwrap_or(false) {
            N::Fg(NamedColor::Grey.into(), vec![N::text("eliminated")])
        } else if pub_state.protected.get(p).copied().unwrap_or(false) {
            N::Bold(vec![N::Fg(
                NamedColor::Foreground.into(),
                vec![N::text("protected")],
            )])
        } else {
            N::Bold(vec![N::Fg(
                NamedColor::Green.into(),
                vec![N::text("active")],
            )])
        };
        let mut discards: Vec<N> = vec![];
        for (i, &c) in pub_state
            .discards
            .get(p)
            .map(Vec::as_slice)
            .unwrap_or(&[])
            .iter()
            .enumerate()
        {
            if i > 0 {
                discards.push(N::text("  "));
            }
            discards.push(card(c));
        }
        table.push(vec![
            (A::Left, vec![N::Player(p)]),
            (
                A::Center,
                vec![N::Bold(vec![N::text(format!(
                    "{}",
                    pub_state.player_points.get(p).copied().unwrap_or(0)
                ))])],
            ),
            (A::Center, vec![status]),
            (A::Left, discards),
        ]);
    }
    table
}

fn help_table() -> Vec<Row> {
    let mut table: Vec<Row> = vec![
        vec![],
        vec![],
        vec![
            (A::Left, vec![N::Bold(vec![N::text("Card")])]),
            (A::Left, vec![N::Bold(vec![N::text("#")])]),
            (A::Left, vec![N::Bold(vec![N::text("Description")])]),
        ],
    ];
    for c in princess_to_guard() {
        table.push(vec![
            (A::Left, vec![card(c)]),
            (A::Left, vec![N::text(format!("{}", deck_count(c)))]),
            (
                A::Left,
                vec![N::Fg(NamedColor::Grey.into(), vec![N::text(c.text())])],
            ),
        ]);
    }
    table
}

impl Renderer for PubState {
    fn render(&self) -> Vec<N> {
        render(self, None, None)
    }
}

impl Renderer for PlayerState {
    fn render(&self) -> Vec<N> {
        render(&self.public, Some(self.player), Some(&self.hand))
    }
}

pub fn card(c: Card) -> N {
    N::Bold(vec![N::Fg(
        c.color().into(),
        vec![N::text(format!("{} ({})", c.name(), c.number()))],
    )])
}

pub fn comma_cards(cards: &[Card]) -> Vec<N> {
    let mut output: Vec<N> = vec![];
    for &c in cards {
        if !output.is_empty() {
            output.push(N::text(", "));
        }
        output.push(card(c));
    }
    output
}
